src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+             <> O.help "Target directory to copy the verified Pact database into"
+            ))
+      <*> rocksParser
+      <*> cwvParser
+
+versionFromText :: (HasCallStack) => Text -> ChainwebVersion
+versionFromText t = case fromText @ChainwebVersionName t of
+  Just a -> lookupVersionByName a
+  Nothing -> error $ "Invalid chainweb version name: " ++ Text.unpack t
+
+checkPactDbsExist :: FilePath -> [ChainId] -> IO ()
+checkPactDbsExist dbDir cids = pooledFor cids $ \cid -> do
+  e <- doesFileExist (chainwebDbFilePath cid dbDir)
+  when (not e) $ do
+    error $ "Pact database doesn't exist for expected chain id " <> Text.unpack (chainIdToText cid)
+
+chainwebDbFilePath :: ChainId -> FilePath -> FilePath
+chainwebDbFilePath cid dbDir =
+  let fileName = mconcat
+        [ "pact-v1-chain-"
+        , Text.unpack (chainIdToText cid)
+        , ".sqlite"
+        ]
+  in dbDir </> fileName
+
+cwvParser :: Parser ChainwebVersion
+cwvParser = fmap versionFromText $ O.strOption
+  (O.long "graph-version"
+    <> O.short 'v'
+    <> O.metavar "CHAINWEB_VERSION"
+    <> O.help "Chainweb version for graph. Only needed for non-standard graphs."
+    <> O.value (toText (_versionName mainnet))
+    <> O.showDefault
+  )
+
+rocksParser :: Parser FilePath
+rocksParser = O.strOption
+  (O.long "rocksdb-dir"
+   <> O.metavar "ROCKSDB DIR"
+   <> O.help "Path to RocksDB directory"
+  )
+
+targetsParser :: Parser BlockHeightTargets
+targetsParser =
+  let
+    makeTargeted = \case
+      [] -> LatestAll
+      ts -> TargetAll (Set.fromList ts)
+    p = fmap makeTargeted $ many $ fmap BlockHeight $ O.option O.auto
+          (O.long "target-blockheight"
+            <> O.short 'b'
+            <> O.metavar "BLOCKHEIGHT"
+            <> O.help "BlockHeight to verify."
+          )
+  in
+  p <|> pure LatestAll
+
+hex :: ByteString -> Text
+hex = Text.decodeUtf8 . Base16.encode
+
+pooledFor :: (Foldable t) => t a -> (a -> IO b) -> IO ()
+pooledFor stuff f = pooledMapConcurrentlyN_ 4 f stuff
+
+allChains :: ChainwebVersion -> [ChainId]
+allChains v = List.sort $ F.toList $ chainIdsAt v (BlockHeight maxBound)
+
+grandsToJson :: [(BlockHeight, Map ChainId (ByteString, BlockHeader))] -> BL.ByteString
+grandsToJson chainHashes =
+  J.encode $ J.Object $ flip List.map chainHashes $ \(height, hashes) ->
+    let key = Text.pack $ show height
+        val = J.Object $ flip List.map (Map.toAscList hashes) $ \(cid, (hash, header)) ->
+                let o = J.Object
+                      [ "hash" J..= hex hash
+                      , "header" J..= J.encodeWithAeson header
+                      ]
+                in (chainIdToText cid, o)
+    in (key J..= val)
+
+-- | Output a Haskell module with the embedded hashes. This module produced
+--   pact-calc, and embedded into the chainweb-node library.
+--
+--   The implementation is a little janky, but it works.
+chainHashesToModule :: [(BlockHeight, Map ChainId (ByteString, BlockHeader))] -> String
```

Comment:
this is the jankiest code in the module. it works, though. it's hard to refactor and it's smelly. if anyone has any suggestions for improvements, please say them.